_source/_posts/2022-04-04-java-cli-device-grant.adoc

```diff
@@ -0,0 +1,412 @@
+
+Create a new `DeviceGrant.java` and add `jackson-databind` as a dependency:
+
+====
+[source,java]
+----
+//DEPS com.fasterxml.jackson.core:jackson-databind:2.13.2 // <.>
+
+import com.fasterxml.jackson.databind.ObjectMapper; // <.>
+
+import java.awt.*; // <.>
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.net.URI;
+import java.net.URLEncoder;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.nio.charset.StandardCharsets;
+import java.time.Duration;
+import java.util.Map;
+
+import static java.util.stream.Collectors.joining;
+
+public class DeviceGrant {
+
+    public static void main(String[] args) throws Exception {
+
+    }
+}
+----
+
+<.> Make sure there is no space between the `//` and `DEPS`.
+<.> Add the following imports, they will be used below.
+<.> AWT? Yes, keep reading I'll get to that 😉
+====
+
+Execute the application from the command line using:
+
+[source,shell]
+----
+jbang DeviceGrant.java
+----
+
+This application doesn't do anything yet, we can fix the in the next few steps!
+
+=== A little more boilerplate
+
+The fluent `HttpClient` is handy, but along with not having JSON support, it doesn't have direct support for form encoded requests. Fortunately, both are easy to fix by adding a couple of helper methods!
+
+In `DeviceGrant.java` add the following two methods:
+
+[source,java]
+----
+// Convert a map into a encoded form body
+private static HttpRequest.BodyPublisher formBody(Map<String, String> params) {
+    // Wrap an encoded String in a BodyPublisher
+    return HttpRequest.BodyPublishers.ofString(
+        // url encode <key>=<value>&
+        params.entrySet().stream()
+            .map(entry -> URLEncoder.encode(entry.getKey(), UTF_8)
+                          + "="
+                          + URLEncoder.encode(entry.getValue(), UTF_8))
+            .collect(joining("&")));
+}
+
+// parse a response body using Jackson
+public static <W> HttpResponse.BodySubscriber<W> parseJson(Class<W> targetType) {
+    return HttpResponse.BodySubscribers.mapping(
+        HttpResponse.BodySubscribers.ofByteArray(),
+        (byte[] bytes) -> {
+            try {
+                ObjectMapper objectMapper = new ObjectMapper();
+                return objectMapper.readValue(bytes, targetType);
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        });
+}
+----
+
+== Initiate the authorization request
+
+From the above link:#setup-okta[Setup Okta] section you have a public "client id" and "issuer" URI, for convenience's sake, set those as `String` variables:
+
+[source,java]
+----
+public static void main(String[] args) throws Exception {
+
+    String clientId = "your-client-id-here" ;
```

Comment:
again nit: why not `var` throughout since java 11+ is mandatory for the post, ignore if its just personal preference as I personally try to show newer java features as much as possible